### PR TITLE
Introduce CONDA_JL_CONDA_EXE environment variable

### PR DESCRIPTION
**Pull request by mkitti**
Here we introduce a new environment variable, `CONDA_JL_CONDA_EXE` which determines the location of the `conda` executable at build time. This in turn is saved in deps.jl as the constant `CONDA_EXE`. This in turn defines the constant `Conda.conda`. If `CONDA_EXE` does not exist due to an old deps.jl file, `Conda.conda` is defined in manner prior to this PR. 

The default `CONDA_EXE` is defined by the following code based on ROOTENV.
```julia
        const CONDA_EXE = if Sys.iswindows()
            p = joinpath(ROOTENV, "Script")
            conda_bat = joinpath(p, "conda.bat")
            isfile(conda_bat) ? conda_bat : joinpath(p, "conda.exe")
        else
            joinpath(ROOTENV, "bin", "conda")
        end
```

Conda.jl also has a method `_install_conda` which installs conda if the executable is not present. To determine the `PREFIX` for this installation the following code is used: `CONDA_EXE_PREFIX = conda |> dirname |> dirname`.

### CODE REVIEW

File: src/Conda.jl
Comment (isuruf) [inline src/Conda.jl:74]: This is impossible. If there's an outdated deps.jl that means there was a update of Conda.jl which means the package has to be built first.

Comment (mkitti) [reply src/Conda.jl:74]: It's possible in a few situations such as using multiple Julia installations with the same `JULIA_DEPOT_PATH` and/or a modified `JULIA_LOAD_PATH`. Julia will only force a rebuild if it is aware that the version changed.

For example, I did this to myself in the course of testing the package by using `pkg"dev Conda"`. In this case, the Conda package was checked out in `~/.julia/dev/Conda` and the deps file at `~/.julia/dev/Conda/deps/deps.jl`was carried over between several versions as I moved between branches or I changed environments.

While we could just throw and error if we detect this condition, encouraging the user to rebuild, just doing the former calculation seems friendlier since we can just follow the old behavior here.

We see this issue with an outdated deps.jl file quite commonly with GR.jl and users will appear in support channels complaining that plotting is broken. The issues were reduced once we just started fixing the issues automatically.

**Comment by mkitti** [comment]
Could we run the CI to see if there are any current issues?

File: src/Conda.jl
Comment (mkitti) [reply src/Conda.jl:74]: We can remove it now, but my recommendation is to keep it for the moment and then remove it at a later time to smooth over the migration process.

File: test/runtests.jl
Comment (tkf) [inline test/runtests.jl:254]: If you are using strange indentation to keep the diff minimal, I suggest `local` (it applies to other places)

```suggestion
            local ROOTENV=joinpath(condadir, "3")
            local CONDA_EXE=default_conda_exe(ROOTENV)
```

File: src/Conda.jl
Comment (tkf) [inline src/Conda.jl:195]: I think this assumption is too strong. It's possible that people use a shim script or symlink. Can you use something like `conda info --json` or `conda run python -c 'import sys; print(sys.executable)'`?

Nitpicks: It's strange that a local variable is all caps. I also notice that src/Conda.jl does not use `|>` for unary function application. I have nothing against `|>` per se, but, I think it's good to follow the style of surrounding code.

Comment (isuruf) [reply src/Conda.jl:195]: @mkitti, why do we need a `CONDA_EXE_PREFIX` when `PREFIX` should work just as fine?

Comment (mkitti) [reply src/Conda.jl:195]: `PREFIX` is determined by `ROOTENV`. https://github.com/JuliaPy/Conda.jl/blob/f1690184f15c875100e35dde187f3dd2a1235db8/src/Conda.jl#L47

We reach this line of code when the conda executable does not exist at `Conda.conda`.
Thus, the objective here is to install the conda executable at the location specified by `CONDA_JL_CONDA_EXE` / `CONDA_EXE` / `Conda.conda`.

Comment (isuruf) [reply src/Conda.jl:195]: You mean when the user sets `CONDA_JL_CONDA_EXE` env variable to a non-existent conda executable? In that case, we should just error in `deps.jl`.

Comment (mkitti) [reply src/Conda.jl:195]: We can add a check at build time. What should we do at runtime when `_install_conda` is invoked?

Comment (isuruf) [reply src/Conda.jl:195]: > We can add a check at build time.

Thanks

> What should we do at runtime when _install_conda is invoked?

Nothing. We just use PREFIX as the if condition above is invoked only if `CONDA_JL_CONDA_EXE` is not set.

Comment (mkitti) [reply src/Conda.jl:195]: Is it only invoked if `CONDA_JL_CONDA_EXE` is not set? It might also be invoked if `CONDA_EXE` gets deleted between build time and run time.

Comment (tkf) [reply src/Conda.jl:195]: I forgot to mention another point: does Conda.jl still use `PREFIX`? If not, it's better to remove it or at least deprecate it.

Comment (mkitti) [reply src/Conda.jl:195]: There's an internal constant called `PREFIX`, but it does not use an environmental variable called PREFIX.

Comment (isuruf) [reply src/Conda.jl:195]: > Is it only invoked if CONDA_JL_CONDA_EXE is not set? It might also be invoked if CONDA_EXE gets deleted between build time and run time.

True. I'm not sure how to handle it. If `CONDA_EXE` gets deleted between build time and run time, I'm not sure if we should install it there.

File: test/runtests.jl
Comment (tkf) [inline test/runtests.jl:161]:
```suggestion
            p = joinpath(ROOTENV, "Scripts")
```

**Review by tkf — COMMENTED** [review]
LGTM! I think it's ready to merge after `CONDA_JL_CONDA_EXE` is documented in README and @stevengj looked at it.

**Comment by mkitti** [comment]
Great. Thanks for the help. It looks like this can also be used to use `mamba` in place of `conda`.

```
$ CONDA_JL_CONDA_EXE=~/mambaforge/bin/mamba julia

(@v1.7) pkg> build Conda
    Building Conda → `~/.julia/dev/Conda/deps/build.log`
Precompiling project...
  1 dependency successfully precompiled in 1 seconds (16 already precompiled)

julia> Conda.conda
"/home/mkitti/mambaforge/bin/mamba"

julia> Conda.list()
[ Info: Running `conda list` in root environment
# packages in environment at /home/mkitti/.julia/conda/3:
```

**Review by tkf — APPROVED** [review]

**Comment by mkitti** [comment]
@stevengj , would you like to review before we merge this?

**Comment by mkitti** [comment]
Would anyone object to merging by the end of this week if we do not hearing anything else?

File: deps/build.jl
Comment (mkitti) [reply deps/build.jl]: Thanks. I made this change in 84c662c

**Comment by mkitti** [comment]
Thank you for the reviews. Can we merge this?

**Comment by mkitti** [comment]
Thanks again for studying this pull request. I would like to merge this tomorrow if I there are no further concerns.

**Pull request merged**
